### PR TITLE
Add referential info

### PR DESCRIPTION
**Pull request by JGSweets**
in this pr:
* adds the referential table
* adds indexes for the table
* updates the parse script to add references
* updates the patent table to include missing application data
* updates the README for the config
* adds a script for setting up the PSQL database

### CODE REVIEW

File: parse_uspto_xml/parse_patent.py [hunk 1]
```diff
@@ -19,11 +22,11 @@
...
 
 
-def get_filenames_from_dir(dirpaths: list):
+def get_filenames_from_dir(dirpaths: list | str):
     """Get filenames from directory"""
 
-    if len(dirpaths) == 1 and not dirpaths[0][-1] == '/':
-        return dirpaths
+    if isinstance(dirpaths, str):
+        dirpaths = [dirpaths]
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:29]: bug fix

File: parse_uspto_xml/parse_patent.py [hunk 2]
```diff
@@ -37,7 +40,7 @@ def get_filenames_from_dir(dirpaths: list):
                     dir_filenames = get_filenames_from_dir([fullpath])
                 filenames += dir_filenames
         else:
-            filenames += dirpath
+            filenames += [dirpath]
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:43]: bug fix

File: parse_uspto_xml/parse_patent.py [hunk 3]
```diff
@@ -46,10 +49,128 @@ def parse_uspto_file(bs, keep_log: bool = False):
     Parses a USPTO patent in a BeautifulSoup object.
     """
 
+    grant_date = bs.get("date-produced", None)
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:52]: new params, not 100% confident on this: needs review @lettergram

File: parse_uspto_xml/parse_patent.py [hunk 4]
```diff
@@ -46,10 +49,128 @@ def parse_uspto_file(bs, keep_log: bool = False):
...
+
     publication_title = bs.find('invention-title').text
     publication_num = bs['file'].split("-")[0]
     publication_date = bs.find('publication-reference').find('date').text
-    application_type = bs.find('application-reference')['appl-type']
+    application_ref_bs = bs.find('application-reference')
+    application_type = application_ref_bs['appl-type']
+    application_date = application_ref_bs.find('date').text
+    application_num = application_ref_bs.find('doc-number').text
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:60]: new

File: parse_uspto_xml/parse_patent.py [hunk 5]
```diff
@@ -46,10 +49,128 @@ def parse_uspto_file(bs, keep_log: bool = False):
...
+
+    referential_documents = []
+    # {uspto_patents.publication_number,reference,cited_by_examiner,document_type,country,metadata (JSON)
+
+    related_docs_bs = bs.find("us-related-documents")
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:65]: new

File: parse_uspto_xml/parse_patent.py [hunk 6]
```diff
@@ -46,10 +49,128 @@ def parse_uspto_file(bs, keep_log: bool = False):
...
+            related_doc["cited_by_examiner"] = False
+            related_doc["reference"] = related_doc_bs.find("doc-number").text
+            related_doc["country"] = related_doc_bs.find("country").text
+            related_doc["metadata"]["kind"] = related_doc_bs.find("kind").text
+            related_doc["metadata"]["date"] = related_doc_bs.find("date").text
+        else:
+            raise KeyError(f"'{related_doc_bs.name}' is not setup to be included in referential documents.")
+        referential_documents.append(related_doc)
+
+    references = []
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:111]: new

File: parse_uspto_xml/parse_patent.py [hunk 7]
```diff
@@ -46,10 +49,128 @@ def parse_uspto_file(bs, keep_log: bool = False):
...
+                    "reference": ref_bs.find("othercit").text,
+                    "cited_by_examiner": "examiner" in ref_bs.find("category").text,
+                    "document_type": "other-reference",
+                    "country": getattr(ref_bs.find("country"), "text", None),
+                    "metadata":{}
+                }
+            references.append(reference)
+        referential_documents += references
+
+    priority_claims = []
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:142]: new

File: parse_uspto_xml/parse_patent.py [hunk 8]
```diff
@@ -46,10 +49,128 @@ def parse_uspto_file(bs, keep_log: bool = False):
...
+                "cited_by_examiner": False,
+                "document_type": "other-reference",
+                "country": getattr(doc_bs.find("country"), "text", None),
+                "metadata":{
+                    "date": getattr(doc_bs.find("date"), "text", None),
+                },
+            })
+        referential_documents += priority_claims
+
+    # check to make sure all keys are proper -- TODO: this should be a test.
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:159]: should be a test, not in the code

File: parse_uspto_xml/parse_patent.py [hunk 9]
```diff
@@ -148,11 +269,15 @@ def build_org(bs_el):
         "publication_title": publication_title,
         "publication_number": publication_num,
         "publication_date": publication_date,
+        "grant_date": grant_date,
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:276]: new

File: parse_uspto_xml/parse_patent.py [hunk 10]
```diff
@@ -244,6 +369,9 @@ def write_to_db(uspto_patent, db=None):
         uspto_patent['publication_number'],
         uspto_patent['publication_date'],
         uspto_patent['application_type'],
+        uspto_patent['grant_date'],
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:376]: updated

File: parse_uspto_xml/parse_patent.py [hunk 11]
```diff
@@ -264,6 +392,9 @@ def write_to_db(uspto_patent, db=None):
         uspto_patent['publication_title'],
         uspto_patent['publication_date'],
         uspto_patent['application_type'],
+        uspto_patent['grant_date'],
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:399]: updated

File: parse_uspto_xml/parse_patent.py [hunk 12]
```diff
@@ -309,7 +444,67 @@ def write_to_db(uspto_patent, db=None):
                           + "section_class_subclass_groups=%s, "
                           + "abstract=%s, description=%s, "
                           + "claims=%s, updated_at=%s", uspto_db_entry)
+        logger.debug(f"DB UPSERT message: {db_cursor.statusmessage}")
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:451]: debug statement

File: parse_uspto_xml/parse_patent.py [hunk 13]
```diff
@@ -407,15 +604,17 @@ def load_local_files(
         logger.error("\n\nErrors\n------------------------\n")
         for e in errors:
             logger.error(e)
-    logger.info(f"\n\nSuccess Count: {success_count}")
+    logger.info("=" * 50)
+    logger.info("=" * 50)
+    logger.info(f"Success Count: {success_count}")
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:618]: better format

File: parse_uspto_xml/parse_patent.py [hunk 14]
```diff
@@ -385,13 +579,16 @@ def load_local_files(
...
+                    write_referential_documents_to_db(
+                        uspto_patent["referential_documents"], db=push_to
+                    )
                 success_count += 1
                 file_success_count += 1
             except Exception as e:
                 exception_tuple = (count, title, e, traceback.format_exc())
                 errors.append(exception_tuple)
-                logger.error(f"Error: {exception_tuple}")
+                logger.error(f"Error: {exception_tuple}", e)
```
Comment (JGSweets) [inline parse_uspto_xml/parse_patent.py:600]: add traceback

File: parse_uspto_xml/utils/db_interface.py [hunk 15]
```diff
@@ -5,10 +5,14 @@
 # load the psycopg to connect to postgresql
 import psycopg2
 import psycopg2.extras
+from dotenv import load_dotenv
```
Comment (JGSweets) [inline parse_uspto_xml/utils/db_interface.py:8]: use `.env` if desired

File: requirements.txt [hunk 16]
```diff
@@ -1,2 +1,4 @@
 bs4
 psycopg2-binary
+lxml
```
Comment (JGSweets) [inline requirements.txt:3]: xml parsing

**Comment by JGSweets** [comment]
Conflict resolution tested on a local PSQL 14 DB.

File: config/psql_setup_script.sh [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        --FOR PSQL >= 15 which allows null not distinct
+        --CREATE UNIQUE INDEX IF NOT EXISTS patent_reference_constraint on uspto_referential_documents
+        --    (uspto_publication_number, reference, document_type, country, (metadata->>'kind')) NULLS NOT DISTINCT;
+        RAISE WARNING "UNCOMMENT THIS SECTION FOR PSQL>=15; NO INDEXES APPLIED"
+    ELSE
+        --FOR PSQL < 15
+        CREATE UNIQUE INDEX IF NOT EXISTS patent_reference_constraint_null on uspto_referential_documents
+            (uspto_publication_number, COALESCE(reference, ''), document_type, COALESCE(country, '')) WHERE (metadata->>'kind') is NULL;
+        CREATE UNIQUE INDEX IF NOT EXISTS patent_reference_constraint_not_null on uspto_referential_documents
+            (uspto_publication_number, COALESCE(reference, ''), document_type, COALESCE(country, ''), (metadata->>'kind')) WHERE (metadata->>'kind') is NOT NULL;
```
Comment (JGSweets) [inline config/psql_setup_script.sh]: null conflict resolution for referential db

File: config/README.md [hunk 18]
```diff
@@ -54,6 +106,25 @@ At this point the script(s) in this repository will work. However, if we wish to
...
+        --FOR PSQL >= 15 which allows null not distinct
+        --CREATE UNIQUE INDEX IF NOT EXISTS patent_reference_constraint on uspto_referential_documents
+        --    (uspto_publication_number, reference, document_type, country, (metadata->>'kind')) NULLS NOT DISTINCT;
+        RAISE WARNING "UNCOMMENT THIS SECTION FOR PSQL>=15; NO INDEXES APPLIED"
+    ELSE
+        --FOR PSQL < 15
+        CREATE UNIQUE INDEX IF NOT EXISTS patent_reference_constraint_null on uspto_referential_documents
+            (uspto_publication_number, COALESCE(reference, ''), document_type, COALESCE(country, '')) WHERE (metadata->>'kind') is NULL;
+        CREATE UNIQUE INDEX IF NOT EXISTS patent_reference_constraint_not_null on uspto_referential_documents
+            (uspto_publication_number, COALESCE(reference, ''), document_type, COALESCE(country, ''), (metadata->>'kind')) WHERE (metadata->>'kind') is NOT NULL;
```
Comment (lettergram) [inline config/README.md]: I'd prefer if we don't create a unique index on a JSON; it creates a very messy index that will be slower and harder to maintain

Comment (JGSweets) [reply config/README.md]: fixed

**Comment by JGSweets** [comment]
Conflict resolution validated on a local PSQL 14 DB for kind as a column.

**Review by lettergram — APPROVED** [review]

**Pull request merged**